File: 0001-0500/0014_longestCommonPrefix.py
class Solution:
    def longestCommonPrefix(self, strs):
        output = strs[0]
        for i in range(1, len(strs)):
            j = 0
            while j < len(strs[i]) and j < len(output):
                if output[j] == strs[i][j]:
                    j += 1
                else:
                    break
            output = output[:j]
        return output

# Alternative solution

class TrieNode:
    def __init__(self):
        self.children = {}
        self.isEnd = False

class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for i, char in enumerate(word):
            if node.isEnd:
                return word[:i]
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.isEnd = True

    def prefix(self, word, answer):
        node = self.root
        for i, char in enumerate(word):
            if node.isEnd:
                return word[:i]
            if char not in node.children:
                node.isEnd = True
                return word[:i]
            node = node.children[char]
        node.isEnd = True
        return word 

class Solution:
    def longestCommonPrefix(self, strs: List[str]) -> str:
        trie = Trie()
        trie.insert(strs[0])
        answer = strs[0]
        for i in range(1, len(strs)):
            answer = trie.prefix(strs[i], answer)
        return answer

File: 0001-0500/0014_longestCommonPrefix.java
class Node {
    HashMap<Character, Node> children;
    boolean isEnd;

    public Node() {
        this.children = new HashMap<>();
        this.isEnd = false;
    }
}

class Trie {
    Node root;

    public Trie() {
        this.root = new Node();
    }

    public void insert(String word) {
        Node node = this.root;

        for (int i = 0; i < word.length(); i++) {
            char c = word.charAt(i);

            if (!node.children.containsKey(c)) {
                node.children.put(c, new Node());
            }

            node = node.children.get(c);
        }

        node.isEnd = true;
    }
}

class Solution {
    public String longestCommonPrefix(String[] strs) {
        Trie trie = new Trie();

        for(String str: strs) {
            if (str.isEmpty()) return "";
            trie.insert(str);
        }

        Node node = trie.root;
        StringBuilder result = new StringBuilder();

        while (true) {
            if (node.children.size() == 1) {
                char key = '*';
                for (char c: node.children.keySet()) {
                    key = c;
                    result.append(key);
                }
                node = node.children.get(key);
                if (node.isEnd) {
                    return result.toString();
                }
            } else {
                break; 
            }
        }
        return result.toString();
    }
}
